### PR TITLE
Wrap single observer to a separate expression

### LINKED ISSUE #1076: Restructure Expression objects
This issue proposes restructuring the internal data structures supporting `Expression` introduced in #1067.  As it stands, this should not have any impact on the user-facing API.

An initial attempt has been made in this commit: 5657dd4b9ec62c1fe2258cb6c0229e4a8b7c0ec8
For future proofing (as I don't trust GH not to garbage collect this), here is the patch created from this commit: 
[0001-Restructuring-Expression-not-yet-tested-with-recursi.patch.txt](https://github.com/enthought/traits/files/4615789/0001-Restructuring-Expression-not-yet-tested-with-recursi.patch.txt)

While this new structure seems to work now, it has not been tested for future proofing with recursion support. Recursion is one that has profound impact on the data structure, and therefore it would be worth testing the proposed structure with recursion, unless it is decided that recursion will never be supported.

For reference, `on_trait_change` supports `"*"` for recursion support, namely `"root.node*.value"` will match `root.node.value`, `root.node.node.value`, and so on. It was decided that as a first cut implementation, `observe` will not support this feature.

### PR DESCRIPTION
**Pull request by kitchoi**
Closes #1076

This PR moves the construction of a single observer to a separate expression object.
`Expression` remains to be user-facing: it aims at providing a user-friendly interface, e.g. instance methods for composing expressions.

The internal expression objects are hidden from the user, and all implement a common interface called `_IExpression`. (That interface is rather basic at the moment). Despite the name, `Expression` is not an instance of `_IExpression`.  An alternative naming suggestion (for either renaming `Expression` or renaming `_IExpression` is very welcome).

**Checklist**
- [x] Tests
- ~Update API reference (`docs/source/traits_api_reference`)~
- ~Update User manual (`docs/source/traits_user_manual`)~
- ~Update type annotation hints in `traits-stubs`~

### CODE REVIEW

File: traits/observers/expression.py [hunk 1]
```diff
@@ -135,162 +132,94 @@ def _as_graphs(self):
...
 
+    def create_graphs(self, branches):
+        return [
+            _ObserverGraph(node=self.observer, children=branches),
+        ]
 
-# _SeriesExpression and _ParallelExpression share an undeclared interface
-# which require the classes to have implemented ``copy`` and ``_create_graphs``
 
+@_IExpression.register
```
Comment (kitchoi) [inline traits/observers/expression.py:180]: Again, nothing does `issubclass` or `isinstance` check. The registeration is unused. The registration is only here for code readability and discovery (e.g. it is easier to see what implement the interface).

**Comment by mdickinson** [comment]
Please can you give some examples of how a user might use `Expression`?

**Comment by kitchoi** [comment]
Sure.

A user will typically create an instance of `Expression` using one of the top-level functions such as `trait`.
e.g. `trait("name")` returns an instance of `Expression`.

The `parse` function from `parsing` can create an instance of `Expression` from a string:
e.g. `parse("a.b.c")` returns an instance of `Expression`.

Then they can extend it using the instance methods on `Expression`. e.g. `trait("name").trait("child")` returns a new instance of `Expression`.

With that, I don't expect `Expression.__init__` to be called directly by the users because there are more convenient ways. That said, the user can, if they *really* wanted to, instantiate an empty expression with `Expression()`, that is fine. They could create one with the internal object `Expression(_expression=...)`, but the naming of the internal variable and docstring hopefully make it clear that such usage is not part of the public interface so breakage is possible.

**Comment by mdickinson** [comment]
Right, but we're specifically making `Expression` user-facing. I'm wondering what a user would use it for.

I don't think we actually need the `Expression` class at all.

File: traits/observers/expression.py [hunk 2]
```diff
@@ -135,162 +132,94 @@ def _as_graphs(self):
...
-        expression._levels = self._levels.copy()
-        if self._prior_expression is not None:
-            expression._prior_expression = self._prior_expression._copy()
-        return expression
+        raise NotImplementedError("'create_graphs' must be implemented.")
 
 
-def _create_graphs(expression, graphs=None):
-    """ Create ObserverGraphs from a given expression.
+@_IExpression.register
```
Comment (mdickinson) [inline traits/observers/expression.py:157]: I think we should be able to get rid of this class entirely.

**Comment by kitchoi** [comment]
> Please can you give some examples of how a user might use Expression?

Ah, you are asking about "using" the expression. I am sorry, I misunderstood your question earlier.
An expression is to be used together with `HasTraits.observe`, e.g.
```
@observe(trait("parent").trait("child"))
def updated(self, event):
    pass
```

Alternatively we could just let `trait("name")` return an instance of `_SingleObserveExpression` (renamed it to `SingleObserverExpression`). Likewise, `trait("name").trait("attr")` will return an instance of `_SeriesExpression` (renamed to `SeriesExpression`).  Then all the instance methods, e.g. `then`, will need to be inherited by `SingleObserverExpression` and `SeriesExpression` and so forth. Users reading the API documentation will need to find the method on the type of expressions being created.  If we wrap all these in a thin facade, then there is only one class users need to go to in order to find documentation.

**Comment by mdickinson** [comment]
> Alternatively we could just let `trait("name")` return an instance of `_SingleObserveExpression` (renamed it to `SingleObserverExpression`). [...]

Yes, I think this is the way to go.

**Comment by kitchoi** [comment]
Currently, the user manual can just point to this `Expression` class and cross-reference its API documentation.  If we are to remove this, all references to `Expression` will refer to an interface, not just docstrings in `observe`, but user manuals. 

While experienced developers can understand interfaces and therefore the `Expression` would seem redundant, the absence of this facade increases the learning curve for those who are new to Python and/or object oriented programming. Considering the goal is to minimize the cognitive load for these users, would you reconsider keeping this facade for that?

**Comment by mdickinson** [comment]
I think the documentation concerns are secondary here. We should start with a simple, clean object structure, and then we can figure out how to best document it. Complicating the code to simplify the docs would be counterproductive.

The way I'd do this would be to have an `Expression` base class (technically an abstract base class, but it doesn't need to be actually defined as inheriting from `abc.ABC`) containing the common code (e.g., the implementation of the `then` method), and have the concrete subclasses like `SingleObserverExpression` inherit from `Expression`. Do away with the interface entirely: just fold the methods into the `Expression` abstract base class.

**Comment by kitchoi** [comment]
No problem. I understand exactly what you meant, and have made exact the same changes you have in #1094 (to the point they are identical).  Would you like me to push those changes here from your PR?

**Comment by mdickinson** [comment]
No,  ignore my PR. I'll close it. :-)

File: traits/observers/expression.py [hunk 3]
```diff
@@ -25,25 +27,8 @@ class Expression:
     ``HasTraits.observe`` method or the ``observe`` decorator.
 
     An Expression is typically created using one of the top-level functions
-    provided in this module, e.g.``trait``.
+    provided in this module, e.g. ``trait``.
```
Comment (kitchoi) [inline traits/observers/expression.py:30]: Drive-by: Without the space Sphinx refuses to display `trait` as code.

**Comment by kitchoi** [comment]
> No, ignore my PR. I'll close it. :-)

Thank you for diving in though. Much appreciated.

File: traits/observers/expression.py [hunk 4]
```diff
@@ -73,10 +58,8 @@ def __or__(self, expression):
         new_expression : traits.observers.expression.Expression
         """
         if self == expression:
-            return self._copy()
-        new = Expression()
-        new._prior_expression = _ParallelExpression([self, expression])
-        return new
+            return self
```
Comment (mdickinson) [inline traits/observers/expression.py:61]: Could we drop this branch? Is it required for correctness, or is it more for optimization?

Comment (mdickinson) [reply traits/observers/expression.py:61]: (And if it's required for correctness, what happens with something like `expr1 | expr2 | expr1`?)

Comment (kitchoi) [reply traits/observers/expression.py:61]: This is an optimization and yes we can drop it.
